paz/models/detection/frustum_pointnet.py

```diff
@@ -0,0 +1,370 @@
+    IntermediateOutputs['heading_scores'] = heading_scores  # BxNUM_HEADING_BIN
+    IntermediateOutputs['heading_residuals_normalized'] = \
+        heading_residuals_normalized  # BxNUM_HEADING_BIN (-1 to 1)
+    IntermediateOutputs['heading_residuals'] = \
+        heading_residuals_normalized * (np.pi / NUM_HEADING_BIN)
+
+    size_scores = tf.slice(output, [0, 3 + NUM_HEADING_BIN * 2],
+                           [-1, NUM_SIZE_CLUSTER])  # BxNUM_SIZE_CLUSTER
+    size_residuals_normalized = tf.slice(output,
+                                         [0,
+                                          3 + NUM_HEADING_BIN * 2 +
+                                          NUM_SIZE_CLUSTER],
+                                         [-1, NUM_SIZE_CLUSTER * 3])
+
+    size_residuals_normalized = tf.reshape(size_residuals_normalized,
+                                           [batch_size, NUM_SIZE_CLUSTER,
+                                            3])  # BxNUM_SIZE_CLUSTERx3
+    IntermediateOutputs['size_scores'] = size_scores
+    IntermediateOutputs['size_residuals_normalized'] = size_residuals_normalized
+    IntermediateOutputs['size_residuals'] = \
+        size_residuals_normalized * tf.expand_dims(
+            tf.constant(g_mean_size_arr,  dtype=tf.float32), 0)
+
+    return IntermediateOutputs
+
+
+def ObjectPointCloudMasking(point_cloud, mask, npoints=512):
+    """ Gather object point clouds according to predicted masks.
+    Input:
+        point_cloud: TF tensor in shape (B,N,C)
+        mask: TF tensor in shape (B,N) of 0 (not pick) or 1 (pick)
+        npoints: int scalar, maximum number of points to keep (default: 512)
+    Output:
+        object_pc: TF tensor in shape (B,npoint,C)
+        indices: TF int tensor in shape (B,npoint,2)
+    """
+
+    def mask_to_indices(mask):
+        indices = np.zeros((mask.shape[0], npoints, 2), dtype=np.int32)
+        for i in range(mask.shape[0]):
+            pos_indices = np.where(mask[i, :] > 0.5)[0]
+            # skip cases when pos_indices is empty
+            if len(pos_indices) > 0:
+                if len(pos_indices) > npoints:
+                    choice = np.random.choice(len(pos_indices),
+                                              npoints, replace=False)
+                else:
+                    choice = np.random.choice(len(pos_indices),
+                                              npoints - len(pos_indices),
+                                              replace=True)
+                    choice = np.concatenate((np.arange(len(pos_indices)),
+                                             choice))
+                np.random.shuffle(choice)
+                indices[i, :, 1] = pos_indices[choice]
+            indices[i, :, 0] = i
+        return indices
+
+    indices = tf.py_function(mask_to_indices, [mask], tf.int32)
+    object_pc = tf.gather_nd(point_cloud, indices)
+    return object_pc, indices
+
+def PointCloudTranslation(point_cloud, logits, IntermediateOutputs,
```

Comment:
function names start with verbs and use lower case. If this is a function used for a tensorflow model, maybe it would be better to use the Layer abstraction